File: baseclient/java/src/main/java/com/aliyun/ccp/baseclient/credentials/AccessTokenCredential.java
package com.aliyun.ccp.baseclient.credentials;

import com.aliyun.ccp.baseclient.utils.TimeUtils;
import com.aliyun.credentials.exception.CredentialException;
import com.aliyun.tea.*;
import com.aliyun.tea.utils.StringUtils;
import com.google.gson.Gson;

import java.io.IOException;
import java.net.URISyntaxException;
import java.security.KeyManagementException;
import java.security.NoSuchAlgorithmException;
import java.text.ParseException;
import java.util.HashMap;
import java.util.Map;

public class AccessTokenCredential {
    private String accessToken;
    private String expireTime;
    private String refreshToken;
    private String domainId;
    private String clientId;
    private String clientSecret;

    public AccessTokenCredential(Map<String, Object> config) {
        this.refreshToken = (String) config.get("refresh_token");
        this.domainId = (String) config.get("domainId");
        this.clientId = (String) config.get("clientId");
        this.clientSecret = (String) config.get("clientSecret");
        this.accessToken = (String) config.get("accessToken");
        this.expireTime = (String) config.get("expireTime");
    }

    public void refreshAccessToken() throws NoSuchAlgorithmException, IOException, KeyManagementException,
            IllegalAccessException, URISyntaxException, TeaException {
        TeaRequest request_ = new TeaRequest();
        request_.protocol = "http";
        request_.method = "POST";
        request_.pathname = "/v2/oauth/token";
        request_.headers = TeaConverter.buildMap(
                new TeaPair("host", this.domainId + ".api.alicloudccp.com"),
                new TeaPair("content-type", "application/x-www-form-urlencoded; charset=utf-8"),
                new TeaPair("date", TimeUtils._getGMTDate()),
                new TeaPair("accept", "application/json"),
                new TeaPair("x-acs-signature-method", "HMAC-SHA1"),
                new TeaPair("x-acs-signature-version", "1.0")
        );
        request_.body = String.format("grant_type=refresh_token&refresh_token=%s&client_id=%s&client_secret=%s",
                this.refreshToken, this.clientId, this.clientSecret);
        TeaResponse response_ = Tea.doAction(request_, new HashMap<>());
        String body = response_.getResponseBody();
        Gson gson = new Gson();
        Map<String, Object> map = gson.fromJson(body, Map.class);
        if (response_.statusCode == 200) {
            this.expireTime = (String) map.get("expire_time");
            this.accessToken = (String) map.get("access_token");
            this.refreshToken = (String) map.get("refresh_token");
        } else {
            throw new TeaException(map);
        }
    }

    public String getAccessToken() throws Exception {
        if (null == this.accessToken) {
            throw new CredentialException("accessToken is null");
        }
        if (shouldRefresh(System.currentTimeMillis())) {
            refreshAccessToken();
        }
        return this.accessToken;
    }

    private boolean shouldRefresh(long now) throws ParseException {
        if (StringUtils.isEmpty(this.expireTime) || StringUtils.isEmpty(this.refreshToken)) {
            return false;
        }
        return now >= (TimeUtils._getTime(this.expireTime) - 5000);
    }

    public void setAccessToken(String accessToken) {
        this.accessToken = accessToken;
    }

    public String getExpireTime() {
        return expireTime;
    }

    public void setExpireTime(String expireTime) {
        this.expireTime = expireTime;
    }

    public String getRefreshToken() {
        return refreshToken;
    }

    public void setRefreshToken(String refreshToken) {
        this.refreshToken = refreshToken;
    }
}

File: baseclient/csharp/core/Credentials/AccessTokenCredential.cs
using System;
using System.Collections.Generic;

using Aliyun.SDK.CCP.Utils;

using Newtonsoft.Json;

using Tea;

namespace Aliyun.SDK.CCP.Credentials
{
    public class AccessTokenCredential
    {
        private string accessToken;
        private string expireTime;
        private string refreshToken;
        private string domainId;
        private string clientId;
        private string clientSecret;
        private string endpoint;

        public AccessTokenCredential(Dictionary<string, object> config)
        {
            this.refreshToken = DictUtils.GetDicValue(config, "RefreshToken").ToSafeString();
            this.domainId = DictUtils.GetDicValue(config, "DomainId").ToSafeString();
            this.endpoint = DictUtils.GetDicValue(config, "Endpoint").ToSafeString();
            this.clientId = DictUtils.GetDicValue(config, "ClientId").ToSafeString();
            this.clientSecret = DictUtils.GetDicValue(config, "ClientSecret").ToSafeString();
            this.accessToken = DictUtils.GetDicValue(config, "AccessToken").ToSafeString();
            this.expireTime = DictUtils.GetDicValue(config, "ExpireTime").ToSafeString();
        }

        public void RefreshAccessToken()
        {
            TeaRequest request = new TeaRequest();
            request.Protocol = "http";
            request.Method = "POST";
            request.Pathname = "/v2/oauth/token";
            request.Headers = new Dictionary<string, string>();
            request.Headers.Add("host", GetHost(this.endpoint, this.domainId + ".api.alicloudccp.com"));
            request.Headers.Add("content-type", "application/x-www-form-urlencoded; charset=utf-8");
            request.Headers.Add("date", TimeUtils.GetGMTDate());
            request.Headers.Add("accept", "application/json");
            request.Headers.Add("x-acs-signature-method", "HMAC-SHA1");
            request.Headers.Add("x-acs-signature-version", "1.0");
            request.Body = string.Format("grant_type=refresh_token&refresh_token={0}&client_id={1}&client_secret={2}",
                this.refreshToken, this.clientId, this.clientSecret);
            TeaResponse response = TeaCore.DoAction(request, new Dictionary<string, object>());
            string body = TeaCore.GetResponseBody(response);
            Dictionary<string, object> bodyDict = JsonConvert.DeserializeObject<Dictionary<string, object>>(body);
            if (response.StatusCode == 200)
            {
                this.expireTime = DictUtils.GetDicValue(bodyDict, "expire_time").ToSafeString();
                this.accessToken = DictUtils.GetDicValue(bodyDict, "access_token").ToSafeString();
                this.refreshToken = DictUtils.GetDicValue(bodyDict, "refresh_token").ToSafeString();
            }
        }

        public bool WithShouldRefresh()
        {
            if (string.IsNullOrWhiteSpace(this.refreshToken) || string.IsNullOrWhiteSpace(this.expireTime))
            {
                return false;
            }
            string expireTimeStr = this.expireTime.Replace('T', ' ').Replace('Z', ' ');
            DateTime expireDate = Convert.ToDateTime(expireTimeStr);
            long expireTimeMillis = expireDate.GetTimeMillis();
            return DateTime.UtcNow.GetTimeMillis() >= (expireTimeMillis - 180);
        }

        public string AccessToken
        {
            get
            {
                if (null == this.accessToken || WithShouldRefresh())
                {
                    RefreshAccessToken();
                }
                return this.accessToken;
            }
            set
            {
                this.accessToken = value;
            }
        }

        public string ExpireTime
        {
            get
            {
                return this.expireTime;
            }
            set
            {
                this.expireTime = value;
            }
        }

        public string GetHost(string endpoint, string domain)
        {
            if (!string.IsNullOrWhiteSpace(endpoint))
            {
                return endpoint;
            }
            else
            {
                return domain;
            }
        }
    }
}
